File: seminar-10/topological_sort/java/TopologicalSort.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TopologicalSort {

    private static boolean dfs(int v, List<List<Integer>> graph, int[] color, List<Integer> order) {
        color[v] = 1;

        for (int to : graph.get(v)) {
            if (color[to] == 0) {
                if (!dfs(to, graph, color, order)) {
                    return false;
                }
            } else if (color[to] == 1) {
                return false;
            }
        }

        color[v] = 2;
        order.add(v);
        return true;
    }

    public static List<Integer> topologicalSort(List<List<Integer>> graph) {
        int n = graph.size();
        int[] color = new int[n];
        List<Integer> order = new ArrayList<>();

        for (int v = 0; v < n; v++) {
            if (color[v] == 0) {
                if (!dfs(v, graph, color, order)) {
                    return new ArrayList<>();
                }
            }
        }

        Collections.reverse(order);
        return order;
    }
}

File: seminar-10/topological_sort/python/topological_sort.py
def dfs(v, graph, color, order):
    color[v] = 1

    for to in graph[v]:
        if color[to] == 0:
            if not dfs(to, graph, color, order):
                return False
        elif color[to] == 1:
            return False

    color[v] = 2
    order.append(v)
    return True


def topological_sort(graph):
    """
    graph — список смежности ориентированного графа
    возвращает топологический порядок вершин
    если в графе есть цикл, возвращает пустой список
    """
    n = len(graph)
    color = [0] * n
    order = []

    for v in range(n):
        if color[v] == 0:
            if not dfs(v, graph, color, order):
                return []

    order.reverse()
    return order
